src/librustc_mir/borrow_check/error_reporting.rs

```diff
@@ -198,19 +198,43 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 let place = &self.move_data.move_paths[mpi].place;
 
                 let ty = place.ty(self.mir, self.infcx.tcx).to_ty(self.infcx.tcx);
-                let note_msg = match self.describe_place_with_options(
-                    place,
-                    IncludingDowncast(true),
-                ) {
-                    Some(name) => format!("`{}`", name),
+                let opt_name = self.describe_place_with_options(place, IncludingDowncast(true));
+                let note_msg = match opt_name {
+                    Some(ref name) => format!("`{}`", name),
                     None => "value".to_owned(),
                 };
-
-                err.note(&format!(
-                    "move occurs because {} has type `{}`, \
-                     which does not implement the `Copy` trait",
-                    note_msg, ty
-                ));
+                let mut note = true;
+                for decl in &self.mir.local_decls {
+                    if decl.ty == ty && decl.name.map(|x| x.to_string()) == opt_name {
+                        err.span_label(
+                            decl.source_info.span,
+                            format!(
+                                "move occurs because {} has type `{}`, \
+                                 which does not implement the `Copy` trait",
+                                note_msg, ty,
+                        ));
+                        note = false;
+                        break;
+                    }
+                }
+                if let ty::TyKind::Param(param_ty) = ty.sty {
+                    let tcx = self.infcx.tcx;
+                    let generics = tcx.generics_of(self.mir_def_id);
+                    let def_id = generics.type_param(&param_ty, tcx).def_id;
+                    if let Some(sp) = tcx.hir().span_if_local(def_id) {
+                        err.span_label(
+                            sp,
+                            "consider adding a `Copy` constraint to this type argument",
```

Comment:
did you intend to include the note in this case?

Reply:
@nikomatsakis No, because if the non-`Copy` type is not a type argument then we can't safely suggest adding a type constraint. At best we could suggest giving it a different type or cloning at the use lines (but that could cause logic errors, depending on the code).

Reply:
Oh, rereading that question, did you mean wether I wanted to show the "move occurs because `x` has type `T`" note/label when suggesting "consider adding a `Copy` constraint"? If so, I did, but it might make sense not to... I would like to keep it just for context, as it teaches that type arguments are not `Copy` by default and also points at the appropriate argument.

Reply:
I was intrigued, and this is the output if you assign `let y = x` in the case above, which seems reasonable and useful to me:

```
error[E0382]: use of moved value: `y`
  --> file.rs:28:9
   |
21 |     pub fn baz<T: Foo>(x: T) -> T {
   |                - consider adding a `Copy` constraint to this type argument
22 |     let y = x;
   |         - move occurs because `y` has type `T`, which does not implement the `Copy` trait
23 |         if 0 == 1 {
24 |             bar::bar(y.zero())
   |                      - value moved here
25 |         } else {
26 |             y.zero()
   |             - value moved here
27 |         };
28 |         y.zero()
   |         ^ value used here after move
```